src/cascadia/TerminalApp/TerminalPage.cpp

```diff
@@ -3001,7 +3002,48 @@ namespace winrt::TerminalApp::implementation
         ShowWindowChanged.raise(*this, args);
     }
 
-    winrt::fire_and_forget TerminalPage::_SearchMissingCommandHandler(const IInspectable /*sender*/, const Microsoft::Terminal::Control::SearchMissingCommandEventArgs args)
+    Windows::Foundation::IAsyncOperation<IVectorView<MatchResult>> TerminalPage::_FindPackagesInCatalogAsync(PackageCatalog catalog, PackageMatchField field, PackageFieldMatchOption matchOption, hstring query)
+    {
+        FindPackagesOptions findPackagesOptions = WindowsPackageManagerFactory::CreateFindPackagesOptions();
+        PackageMatchFilter filter = WindowsPackageManagerFactory::CreatePackageMatchFilter();
+        filter.Field(field);
+        filter.Option(matchOption);
+        filter.Value(query);
+        findPackagesOptions.Filters().Append(filter);
+        findPackagesOptions.ResultLimit(20);
+        FindPackagesResult findPackagesResult{ co_await catalog.FindPackagesAsync(findPackagesOptions) };
+
+        co_return findPackagesResult.Matches();
+    }
+
+    Windows::Foundation::IAsyncOperation<IVectorView<MatchResult>> TerminalPage::_FindPackageAsync(hstring query)
+    {
+        PackageManager packageManager = WindowsPackageManagerFactory::CreatePackageManager();
+        PackageCatalogReference catalogRef{
+            packageManager.GetPredefinedPackageCatalog(PredefinedPackageCatalog::OpenWindowsCatalog)
+        };
+        ConnectResult connectResult = catalogRef.Connect();
+        if (connectResult.Status() != ConnectResultStatus::Ok)
+        {
+            co_return nullptr;
+        }
+        PackageCatalog catalog = connectResult.PackageCatalog();
+
+        auto pkgList = co_await _FindPackagesInCatalogAsync(catalog, PackageMatchField::Command, PackageFieldMatchOption::StartsWithCaseInsensitive, query);
+        if (pkgList.Size() > 0)
+        {
+            co_return pkgList;
+        }
+        pkgList = co_await _FindPackagesInCatalogAsync(catalog, PackageMatchField::Name, PackageFieldMatchOption::ContainsCaseInsensitive, query);
+        if (pkgList.Size() > 0)
+        {
+            co_return pkgList;
+        }
+        pkgList = co_await _FindPackagesInCatalogAsync(catalog, PackageMatchField::Moniker, PackageFieldMatchOption::ContainsCaseInsensitive, query);
+        co_return pkgList;
+    }
```

Comment:
You could consider rewriting this as a loop. Pseudo code:
```cpp
struct Search {
    PackageMatchField field;
    PackageFieldMatchOption matching;
};
static constexpr std::array searches{ ... };

for (const auto& search : searches) {
    const auto matches = co_await ...;
    if (matches.Size() > 0) {
        co_return matches;
    }
}
```

This has the advantage that we can reuse objects across iterations easily:
```cpp
struct Search {
    PackageMatchField field;
    PackageFieldMatchOption option;
};
static constexpr std::array searches{ ... };

const auto filter = WindowsPackageManagerFactory::CreatePackageMatchFilter();
filter.Value(query);

const auto options = WindowsPackageManagerFactory::CreateFindPackagesOptions();
options.Filters().Append(filter);
options.ResultLimit(20);

for (const auto& search : searches) {
    filter.Field(search.field);
    filter.Option(search.option);

    const auto result = co_await catalog.FindPackagesAsync(options);
    const auto matches = result.Matches();

    if (matches.Size() > 0) {
        co_return matches;
    }
}
```

Whether this works or not depends on whether winget is robust against object reuse of course. But I'd expect this to just work.